File: func_handler.py
from typing import Callable

from classes import Point

WIDTH, HEIGHT = 641, 641

assert WIDTH % 2 == 1 and HEIGHT % 2 == 1

fn_type = Callable[[float, float], float]


def calc_fn(fn: fn_type, x: float, y: float) -> float:
    """Calculates a function with the screen x and y as if they were on an axis"""
    return fn(x - WIDTH // 2, y - HEIGHT // 2)


def calc_square(fn: fn_type, x_top: int, y_top: int) -> tuple[float, float, float, float]:
    """Calculate a square using the top left corner while True means green (greater than 1, inside) and false blue (less than 1, outside) and the points are clockwise"""
    return (
        calc_fn(fn, x_top, y_top),
        calc_fn(fn, x_top + 1, y_top),
        calc_fn(fn, x_top + 1, y_top + 1),
        calc_fn(fn, x_top, y_top + 1),
    )


def _generate_point_of_square(x_top: int, y_top: int, idx: int) -> Point:
    """Returns a point based on the idx in square"""
    if idx < 0:
        idx = 4 + idx
    return Point(x_top + int(idx in (1, 2)), y_top + int(idx in (2, 3)))


def _get_line_point(p1: Point, p2: Point, p1_val: float, p2_val: float) -> Point:
    """Returns a point of the contour"""
    k = abs(p1_val - 1)
    l = abs(1 - p2_val)
    return Point(((p1.x * l + p2.x * k) / (k + l)), ((p1.y * l + p2.y * k) / (k + l)))


def get_lines_from_square(square: tuple[float, float, float, float], x_top: int, y_top: int) -> list[
    tuple[Point, Point]]:
    """Returns a list with tuples of start and end of lines to draw"""
    bool_square = [v > 1 for v in square]
    if sum(bool_square) == 4 or sum(bool_square) == 0:
        return []
    elif sum(bool_square) == 1:
        idx = bool_square.index(True)
        p = _generate_point_of_square(x_top, y_top, idx)
        next_idx = (idx + 1) % 4
        next_p = _generate_point_of_square(x_top, y_top, next_idx)
        prev_p = _generate_point_of_square(x_top, y_top, idx - 1)
        return [
            (
                _get_line_point(p, next_p, square[idx], square[next_idx]),
                _get_line_point(prev_p, p, square[idx - 1], square[idx])
            )
        ]
    elif sum(bool_square) == 2:
        idx = bool_square.index(True)
        if bool_square[idx] and bool_square[(idx + 2) % 4]:  # diagonals
            p = _generate_point_of_square(x_top, y_top, idx)
            p2 = _generate_point_of_square(x_top, y_top, (idx + 1) % 4)
            p3 = _generate_point_of_square(x_top, y_top, (idx + 2) % 4)
            t1 = (
                _get_line_point(p, p2, square[idx], square[(idx + 1) % 4]),
                _get_line_point(p2, p3, square[(idx + 1) % 4], square[(idx + 2) % 4])
            )
            p4 = _generate_point_of_square(x_top, y_top, idx - 1)
            t2 = (
                _get_line_point(p3, p4, square[(idx + 2) % 4], square[idx - 1]),
                _get_line_point(p4, p, square[idx - 1], square[idx])
            )
            return [t1, t2]
        else:  # adjacent
            if bool_square[(idx + 1) % 4]:
                idx, next_idx = idx, (idx + 1) % 4
            else:
                idx, next_idx = idx - 1, idx
            next_p = _generate_point_of_square(x_top, y_top, next_idx)
            next_next_p = _generate_point_of_square(x_top, y_top, (next_idx + 1) % 4)
            p = _generate_point_of_square(x_top, y_top, idx)
            prev_p = _generate_point_of_square(x_top, y_top, idx - 1)
            return [
                (
                    _get_line_point(next_p, next_next_p, square[next_idx], square[(next_idx + 1) % 4]),
                    _get_line_point(prev_p, p, square[idx - 1], square[idx])
                )
            ]
    elif sum(bool_square) == 3:
        idx = bool_square.index(False)
        p = _generate_point_of_square(x_top, y_top, idx)
        next_idx = idx + 1 % 4
        next_p = _generate_point_of_square(x_top, y_top, next_idx)
        prev_p = _generate_point_of_square(x_top, y_top, idx - 1)
        return [
            (
                _get_line_point(p, next_p, square[idx], square[next_idx]),
                _get_line_point(prev_p, p, square[idx - 1], square[idx])
            )
        ]

File: ts_version/src/funcHandler.ts
const [WIDTH, HEIGHT] = [1024, (1024 * 9) / 16];

type fnType = (x: number, y: number) => number;

const calcFn = (fn: fnType, x: number, y: number) => {
  try {
    return fn(x - Math.floor(WIDTH / 2), y - Math.floor(HEIGHT / 2));
  } catch (e) {
    return 1;
  }
};

const calcSquare = (fn: fnType, xTop: number, yTop: number) => [
  calcFn(fn, xTop, yTop),
  calcFn(fn, xTop + 1, yTop),
  calcFn(fn, xTop + 1, yTop + 1),
  calcFn(fn, xTop, yTop + 1),
];

const generatePointOfSquare = (xTop: number, yTop: number, idx: number) => {
  return {
    x: xTop + ([1, 2].includes(idx) ? 1 : 0),
    y: yTop + ([2, 3].includes(idx) ? 1 : 0),
  };
};

const getLinePoint = (p1: Point, p2: Point, p1Val: number, p2Val: number) => {
  const k = Math.abs(p1Val - 1);
  const l = Math.abs(1 - p2Val);
  return {
    x: (p1.x * l + p2.x * k) / (k + l),
    y: (p1.y * l + p2.y * k) / (k + l),
  };
};

const getLinesFromSquare = (square: number[], xTop: number, yTop: number) => {
  const newArr = square.map((v) => (v > 1 ? 1 : 0));
  const sum = newArr.reduce((a: number, b: number) => a + b, 0);
  if (sum === 4 || sum === 0) {
    return [];
  } else if (sum === 1) {
    const idx = newArr.indexOf(1);
    const p = generatePointOfSquare(xTop, yTop, idx);
    const nextIdx = (idx + 1) % 4;
    const prevIdx = (idx + 3) % 4;
    const nextP = generatePointOfSquare(xTop, yTop, nextIdx);
    const prevP = generatePointOfSquare(xTop, yTop, prevIdx);
    return [
      [
        getLinePoint(p, nextP, square[idx], square[nextIdx]),
        getLinePoint(prevP, p, square[prevIdx], square[idx]),
      ],
    ];
  } else if (sum === 2) {
    const idx = newArr.indexOf(1);
    if (newArr[(idx + 2) % 4] === 1) {
      const [idx2, idx3, idx4] = [1, 2, 3].map((i) => (idx + i) % 4);
      const [p, p2, p3, p4] = [idx, idx2, idx3, idx4].map((i) =>
        generatePointOfSquare(xTop, yTop, i)
      );
      const t1 = [
        getLinePoint(p, p2, square[idx], square[idx2]),
        getLinePoint(p2, p3, square[idx2], square[idx3]),
      ];
      const t2 = [
        getLinePoint(p3, p4, square[idx3], square[idx4]),
        getLinePoint(p4, p, square[idx4], square[idx]),
      ];
      return [t1, t2];
    } else {
      const [currentIdx, nextIdx] =
        newArr[(idx + 1) % 4] == 1
          ? [idx, (idx + 1) % 4]
          : [(idx + 3) % 4, idx];
      const nextNextIdx = (nextIdx + 1) % 4;
      const prevIdx = (currentIdx + 3) % 4;
      const [prevP, p, nextP, nextNextP] = [
        prevIdx,
        currentIdx,
        nextIdx,
        nextNextIdx,
      ].map((i) => generatePointOfSquare(xTop, yTop, i));
      return [
        [
          getLinePoint(nextP, nextNextP, square[nextIdx], square[nextNextIdx]),
          getLinePoint(prevP, p, square[prevIdx], square[currentIdx]),
        ],
      ];
    }
  } else {
    const idx = newArr.indexOf(0);
    const p = generatePointOfSquare(xTop, yTop, idx);
    const nextIdx = (idx + 1) % 4;
    const prevIdx = (idx + 3) % 4;
    const nextP = generatePointOfSquare(xTop, yTop, nextIdx);
    const prevP = generatePointOfSquare(xTop, yTop, prevIdx);
    return [
      [
        getLinePoint(p, nextP, square[idx], square[nextIdx]),
        getLinePoint(prevP, p, square[prevIdx], square[idx]),
      ],
    ];
  }
};

const getLinesFromFn = (fn: fnType) => {
  const lst: Point[][] = [];
  for (let x = WIDTH - 1; x >= 0; x--) {
    for (let y = HEIGHT - 1; y >= 0; y--) {
      const sqr = calcSquare(fn, x, y);
      const res = getLinesFromSquare(sqr, x, y);
      res.forEach((r) => lst.push(r));
    }
  }
  return lst;
};
